Improvement: DO-3034 support seamless handling of refresh tokens

krzysztof-causalens opened the pull request:
<!--- The title format is expected to follow the pattern:-->
<!--- CHANGE TYPE: Ticket Name -->
<!--- Docs: DO-100 Update PR Template -->
<!--- Where CHANGE TYPEs are: Docs, Breaking, Improvement, Fix, Refactor, Feat, Security -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it has a spec, please link to the spec here. -->

This PR introduces handling of refresh tokens to the framework core. 

When any request fails with a 401 or 403 status code, we initiate a token refresh process. 
If the current auth config supports token refresh, it can choose to use the previous expired token and attached `dara_refresh_token` cookie to produce a new session token.
If successful, the token is replaced in a new global store and the request is retried.
If unsuccessful, the original failed response is returned to the caller.

## Implementation Description
<!--- Why did you follow this implementation approach- -->
<!--- Is there any background knowledge necessary to understand the approach taken- -->
<!--- Describe the limitations, pitfalls and tradeoffs of the approach- -->

### Client-side

Session token has been moved out of a React context into a singleton global store. This gives us more control over refreshing tokens. For example, once a token refresh is already initiated, other attempts to retrieve the token will block and wait for the in-flight refresh to go through. 

The above is achieved utilizing a simple store with a locking mechanism, where in-flight refreshes are represented by promises.

Auth internals like the `useSessionToken` hook have been updated to pull/subscribe to the store instead of the context. 


### Backend

The `refresh-token` endpoint is now included by default in dara core. The base config provides an implementation of a `refresh_token` method which raises a 400 - refresh not supported - error. 

An auth config can be configured to support token refresh by:
- implement the aforementioned method to sign a new token, reusing the previous session_id for continuity
- add some mechanism to set-cookie with the `dara_refresh_token` used, e.g. via custom `components_config` and endpoints such as `sso-callback` for SSO auth

The above is utilized in the internal SSO auth package.  

## Any new dependencies Introduced
<!--- Where there any dependencies added? Why?- -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Unit tests added
Manually tested with internal SSO auth implementation

## PR Checklist:
<!--- Go over all the following points, and ensure it has all been checked with an `x` -->

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [ ] I have added comments to all the bits that are hard to follow.
- [ ] I have added/updated Documentation.
- [ ] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):
<!--- For UI changes make sure to add an image or GIF showcasing the change-->

Roman-Kornev commented on packages/dara-core/js/shared/global-state-store.tsx:
```diff
@@ -0,0 +1,134 @@
+
+        try {
+            result = await fn();
+            this.#state[key] = result;
+
+            // success - notify subscribers, unlock pending promises and set the value
+            this.#notify(key, result);
+            unlock(result);
+            this.setValue(key, result);
```
Isn't this `setValue` doing the same thing as 3 lines above?

krzysztof-causalens replied:
Good catch, yeah we'll notify twice here. I'll remove the explicit one 👍

patricia-causalens commented on packages/dara-components/changelog.md:
```diff
@@ -2,7 +2,7 @@
 ---
 
-# NEXT
+## 1.14.0-alpha.0
```
We should be careful here to not forget to update to the proper version later

krzysztof-causalens replied:
Good catch, I'll reset this to NEXT before merging

patricia-causalens reviewed (approved):
LGTM, just missing changelog entry

sam-causalens commented on packages/dara-core/js/auth/auth-wrapper.tsx:
```diff
@@ -1,16 +1,54 @@
+            }>(token);
+
+            // If it's about to expire within 5 minutes
+            if (decoded.exp * 1000 - Date.now() < 5 * 60 * 1000) {
```
I think 5 minutes is too long. Our target time for the token length is only 10minutes, so this will lead it to refreshing every 5 minutes which seems excessive. I'd be okay with either not having this at all as I think it overcomplicates it for not much benefit (as everything will now block and wait for the refresh anyway?) or reducing this timer down to 10-15 seconds?

krzysztof-causalens replied:
Sure, I agree the way things work now we can probably safely remove it and just let requests fail.

sam-causalens commented on packages/dara-core/js/auth/auth-wrapper.tsx:
```diff
@@ -48,38 +86,61 @@ function DynamicAuthComponent(props: { component: AuthComponent }): JSX.Element
     }
 
+    // check token expiry every minute
+    useEffect(() => {
```
What was the reasoning behind this interval? Why not just wait for a request to fail and then refresh the token?

krzysztof-causalens replied:
To be fair I can't quite recall, I think the idea originated when we were targeting a longer refresh time so we were trying to refresh ahead of time. With the low token time now we can get away without it.

sam-causalens commented on packages/dara-core/js/shared/global-state-store.tsx:
```diff
@@ -0,0 +1,133 @@
+            return this.#locks[key] as Promise<TState[TKey]>;
+        }
+
+        let unlock: (res: TState[TKey]) => void;
```
Can you add some comments in here to explain to someone later what is happening

sam-causalens reviewed (changes requested):
Looks Good Krys, left a couple of comments to think about.

krzysztof-causalens commented:
- removed periodic token refresh, is not necessary with expected refresh token times and is redundant as we now gracefully handle failed requests by retrying them after refreshing
- removed `jwt-decode` dependency as we no longer have to decode the token on the client side to check it's expiry date

krzysztof-causalens commented:
I realized there was an issue where in-progress live WS connections would have stale ContextVar values, such as ID_TOKEN if one is used by an auth config. 

ContextVars can't be updated from the outside, so I've added a new memory object stream into the WS handler object:
- The WS handler logic now checks for pending token data updates in the receive stream and updates the context whenever one is found
- The refresh endpoint sends a token data update into the send stream

Added a test to verify the above behaviour. 

Tagging another alpha version - I'll test how that interacts with multi-tab tomorrow, I have a feeling we might run into issues when two tabs decide to refresh their tokens simultaneously

krzysztof-causalens commented:
Looking more into the cross-tab issues, we've decided to take a different approach.

1) The client-side `GlobalStore` used to store the session token is instead a wrapper around `localStorage`. This means the values are shared cross-tab, and subscriptions use `storage` events so one tab performing a refresh will automatically notify all active tabs.
2) To update live WS connections, the client sends a special `token_update` WS message for the backend to update its in-memory ContextVars.
3) To prevent multiple tabs from performing concurrent token refreshes, the calls to actual token refresh logic are wrapped with a short-lived deduping async cache. Simultaneous attempts to refresh the same token will be deduped and resolve to the same value (due to underlying locks), and subsequent ones within a short window (5s) will also return the same cached value.

Roman-Kornev commented on packages/dara-core/dara/core/auth/utils.py:
```diff
@@ -100,3 +102,120 @@ def get_user_data():
+
+
+async def cached_refresh_token(
+    func: Callable[[TokenData, str], Tuple[str, str]], old_token_data: TokenData, refresh_token: str
```
`func` should probably be something like `do_refresh_token`

sam-causalens reviewed (approved):
This LGTM, thanks for this Krys much appreciated.
